File: intrinsic/assets/dependencies/resolver.cc
// Copyright 2026 Intrinsic Innovation LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "intrinsic/assets/dependencies/resolver.h"

#include <memory>
#include <string>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/functional/function_ref.h"
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "google/protobuf/any.pb.h"
#include "google/protobuf/descriptor.h"
#include "google/protobuf/descriptor.pb.h"
#include "google/protobuf/message_lite.h"
#include "intrinsic/assets/id_utils.h"
#include "intrinsic/assets/interface_utils.h"
#include "intrinsic/assets/proto/field_metadata.pb.h"
#include "intrinsic/assets/proto/v1/dependency.pb.h"
#include "intrinsic/assets/proto/v1/grpc_connection.pb.h"
#include "intrinsic/assets/proto/v1/resolved_dependency.pb.h"
#include "intrinsic/executive/engine/skill_action.h"
#include "intrinsic/util/proto/walk_messages.h"
#include "intrinsic/util/status/status_macros.h"

namespace intrinsic {
namespace assets {

namespace {

absl::Status ProcessField(
    google::protobuf::Message& message,
    const google::protobuf::FieldDescriptor* field,
    const google::protobuf::FieldDescriptor* value_field, /* for map values */
    absl::FunctionRef<absl::Status(google::protobuf::Message&)> process_fn) {
  if (field->is_repeated() && !field->is_map()) {
    for (int i = 0; i < message.GetReflection()->FieldSize(message, field);
         ++i) {
      INTR_RETURN_IF_ERROR(
          process_fn(*message.GetReflection()->MutableRepeatedMessage(
              &message, field, i)));
    }
  } else if (field->is_map()) {
    for (int i = 0; i < message.GetReflection()->FieldSize(message, field);
         ++i) {
      google::protobuf::Message* map_entry =
          message.GetReflection()->MutableRepeatedMessage(&message, field, i);
      INTR_RETURN_IF_ERROR(process_fn(
          *map_entry->GetReflection()->MutableMessage(map_entry, value_field)));
    }
  } else {
    INTR_RETURN_IF_ERROR(
        process_fn(*message.GetReflection()->MutableMessage(&message, field)));
  }
  return absl::OkStatus();
}

}  // namespace

absl::Status Resolver::ResolveAnyField(
    google::protobuf::Message& any_message, executive::SkillAction skill_action,
    const google::protobuf::DescriptorPool* pool,
    google::protobuf::MessageFactory* message_factory) const {
  if (pool == nullptr || message_factory == nullptr) {
    return absl::InvalidArgumentError(
        "Descriptor pool or message factory is null");
  }

  if (any_message.ByteSizeLong() == 0) {
    return absl::OkStatus();
  }

  const google::protobuf::Reflection* reflection = any_message.GetReflection();
  const google::protobuf::Descriptor* any_descriptor =
      any_message.GetDescriptor();
  const google::protobuf::FieldDescriptor* type_url_field =
      any_descriptor->FindFieldByName("type_url");
  const google::protobuf::FieldDescriptor* value_field =
      any_descriptor->FindFieldByName("value");

  if (type_url_field == nullptr ||
      type_url_field->type() !=
          google::protobuf::FieldDescriptor::TYPE_STRING ||
      value_field == nullptr ||
      value_field->type() != google::protobuf::FieldDescriptor::TYPE_BYTES) {
    return absl::InternalError(
        "Message of type google.protobuf.Any does not have type_url or value "
        "fields.");
  }

  std::string type_url = reflection->GetString(any_message, type_url_field);
  if (type_url.empty()) {
    return absl::InvalidArgumentError("Any message has empty type URL.");
  }

  std::string full_type_name;
  if (!google::protobuf::Any::ParseAnyTypeUrl(type_url, &full_type_name)) {
    return absl::InvalidArgumentError(
        absl::StrCat("Could not parse Any type URL: ", type_url));
  }

  const google::protobuf::Descriptor* descriptor =
      pool->FindMessageTypeByName(full_type_name);
  if (descriptor == nullptr) {
    return absl::NotFoundError(
        absl::StrCat("Could not find descriptor for type ", full_type_name));
  }

  const google::protobuf::Message* prototype =
      message_factory->GetPrototype(descriptor);
  if (prototype == nullptr) {
    return absl::InternalError(absl::StrCat(
        "Failed to get prototype for descriptor: ", full_type_name));
  }
  std::unique_ptr<google::protobuf::Message> unpacked_msg(prototype->New());
  if (!unpacked_msg->ParseFromString(
          reflection->GetString(any_message, value_field))) {
    return absl::InvalidArgumentError(
        absl::StrCat("Failed to unpack Any proto with type URL: ", type_url));
  }

  INTR_RETURN_IF_ERROR(RecursivelyWalkMessage(
      *unpacked_msg,
      [this, skill_action, pool, message_factory](
          google::protobuf::Message& msg) -> absl::StatusOr<bool> {
        return this->ResolveDependency(msg, skill_action, pool,
                                       message_factory);
      }));

  std::string new_value;
  if (!unpacked_msg->SerializeToString(&new_value)) {
    return absl::InternalError(
        "Failed to serialize resolved message to string.");
  }
  reflection->SetString(&any_message, value_field, new_value);

  return absl::OkStatus();
}

absl::Status Resolver::ResolveStub(
    ::intrinsic_proto::assets::v1::ResolvedDependency& resolved_dependency,
    const ResolveMessageOptions& options) const {
  const std::string& name = resolved_dependency.name();
  if (name.empty()) {
    return absl::OkStatus();
  }

  for (const auto& interface : options.requires_interfaces) {
    if (interface.starts_with(kGrpcUriPrefix)) {
      ::intrinsic_proto::assets::v1::ResolvedDependency::Interface
          resolved_interface;
      resolved_interface.mutable_grpc()->mutable_connection()->set_address(
          address_);
      auto* metadata = resolved_interface.mutable_grpc()
                           ->mutable_connection()
                           ->mutable_metadata()
                           ->Add();
      metadata->set_key(instance_header_name_);
      metadata->set_value(name);
      (*resolved_dependency.mutable_interfaces())[interface] =
          resolved_interface;
    } else if (interface.starts_with(kDataUriPrefix)) {
      ::intrinsic_proto::assets::v1::ResolvedDependency::Interface
          resolved_interface;
      const auto id_package = assets::PackageFrom(name);
      if (!id_package.ok()) {
        LOG(ERROR) << "Failed to create data asset ID from dependency name: "
                   << id_package.status();
        continue;
      }
      const auto id_name = assets::NameFrom(name);
      if (!id_name.ok()) {
        LOG(ERROR) << "Failed to create data asset ID from dependency name: "
                   << id_name.status();
        continue;
      }

      resolved_interface.mutable_data()->mutable_id()->set_package(*id_package);
      resolved_interface.mutable_data()->mutable_id()->set_name(*id_name);
      (*resolved_dependency.mutable_interfaces())[interface] =
          resolved_interface;
    } else {
      return absl::InvalidArgumentError(
          absl::StrCat("Unsupported dependency type: ", interface));
    }
  }
  if (options.requires_object) {
    resolved_dependency.mutable_object()->set_name(name);
  }

  resolved_dependency.clear_name();
  return absl::OkStatus();
}
absl::Status Resolver::ResolveMessage(
    google::protobuf::Message& message,
    const ResolveMessageOptions& options) const {
  // The message might be a dynamic proto so we serialize and deserialize
  // for modification. Using DynamicCastMessage would be cleaner but does
  // not work (gives nullptr) and static casting produces a segfault.
  std::string serialized = message.SerializeAsString();
  ::intrinsic_proto::assets::v1::ResolvedDependency resolved_dependency;
  if (!resolved_dependency.ParseFromString(serialized)) {
    return absl::InvalidArgumentError(
        "Failed to parse ResolvedDependency from message.");
  }
  INTR_RETURN_IF_ERROR(ResolveStub(resolved_dependency, options));
  // We clear gRPC connections if we are not resolving for execution.
  if (!options.include_connection_info) {
    for (auto& [interface, interface_dependency] :
         *resolved_dependency.mutable_interfaces()) {
      if (interface_dependency.has_grpc()) {
        interface_dependency.mutable_grpc()->clear_connection();
      }
    }
  }
  serialized = resolved_dependency.SerializeAsString();
  if (!message.ParseFromString(serialized)) {
    return absl::InvalidArgumentError(
        "Failed to parse ResolvedDependency back into message.");
  }
  return absl::OkStatus();
}

absl::StatusOr<bool> Resolver::ResolveDependency(
    google::protobuf::Message& message, executive::SkillAction skill_action,
    const google::protobuf::DescriptorPool* pool,
    google::protobuf::MessageFactory* message_factory) const {
  std::vector<const google::protobuf::FieldDescriptor*> fields;
  message.GetReflection()->ListFields(message, &fields);
  for (const google::protobuf::FieldDescriptor* field : fields) {
    // Skip if not a message type.
    if (field->cpp_type() !=
        google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE) {
      continue;
    }

    // Handle Any type.
    const google::protobuf::FieldDescriptor* any_check_field = field;
    bool is_map = field->is_map();
    if (is_map) {
      any_check_field = field->message_type()->FindFieldByName("value");
    }

    if (any_check_field != nullptr &&
        any_check_field->cpp_type() ==
            google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE &&
        any_check_field->message_type()->full_name() ==
            google::protobuf::Any::descriptor()->full_name()) {
      const google::protobuf::FieldDescriptor* value_field =
          is_map ? any_check_field : nullptr;
      const auto status = ProcessField(
          message, field, value_field,
          [this, skill_action, pool,
           message_factory](google::protobuf::Message& msg) -> absl::Status {
            return ResolveAnyField(msg, skill_action, pool, message_factory);
          });
      if (!status.ok()) {
        LOG(ERROR) << "Failed to resolve Any field: " << status
                   << " Skipping dependency resolution for this Any proto.";
      }
      continue;
    }

    // Only process fields that are of type ResolvedDependency or a map with
    // value type ResolvedDependency.
    const google::protobuf::FieldDescriptor* value_field = nullptr;
    if (field->is_map()) {
      value_field = field->message_type()->FindFieldByName("value");
      if (value_field == nullptr ||
          value_field->cpp_type() !=
              google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE ||
          value_field->message_type()->full_name() !=
              intrinsic_proto::assets::v1::ResolvedDependency::descriptor()
                  ->full_name()) {
        continue;
      }
    } else if (field->message_type()->full_name() !=
               intrinsic_proto::assets::v1::ResolvedDependency::descriptor()
                   ->full_name()) {
      continue;
    }

    // Pull out the dependency information from the field metadata. If there is
    // no dependency information, we assume here that the user does not want to
    // treat this field as a dependency specification.
    ResolveMessageOptions resolve_message_options = {};
    const google::protobuf::FieldOptions& options = field->options();
    if (options.HasExtension(intrinsic_proto::assets::field_metadata)) {
      const intrinsic_proto::assets::FieldMetadata& field_metadata =
          options.GetExtension(intrinsic_proto::assets::field_metadata);
      resolve_message_options.requires_interfaces.insert(
          resolve_message_options.requires_interfaces.end(),
          field_metadata.dependency().requires_().begin(),
          field_metadata.dependency().requires_().end());
      resolve_message_options.requires_object =
          field_metadata.dependency().has_requires_object();
      if (field_metadata.dependency().has_skill_annotations()) {
        resolve_message_options.include_connection_info =
            field_metadata.dependency()
                .skill_annotations()
                .always_provide_connection_info();
      }
      if (!field_metadata.has_dependency()) {
        // Since there is no dependency annotation, we assume here that the
        // user does not want to treat this field as a dependency
        // specification.
        continue;
      }
    } else {
      // Since there is no field metadata, we assume here that the user does
      // not want to treat this field as a dependency specification.
      continue;
    }

    // We always include connection info during skill execution.
    if (skill_action == executive::SkillAction::Execution) {
      resolve_message_options.include_connection_info = true;
    }

    // Resolve the dependency information for repeated fields, maps fields, and
    // message fields.
    INTR_RETURN_IF_ERROR(
        ProcessField(message, field, value_field,
                     [this, &resolve_message_options](
                         google::protobuf::Message& msg) -> absl::Status {
                       return ResolveMessage(msg, resolve_message_options);
                     }));
  }

  // Recurse into submessages.
  return true;
}

absl::Status Resolver::PopulateFallbackValues(
    google::protobuf::Message& parameter,
    const absl::flat_hash_map<std::string, std::string>&
        fallback_manifest_dependencies) const {
  // Note that we do not support fallback dependencies for nested fields.
  for (int i = 0; i < parameter.GetDescriptor()->field_count(); ++i) {
    const google::protobuf::FieldDescriptor* field =
        parameter.GetDescriptor()->field(i);
    // Skip if not a message type.
    if (field->cpp_type() !=
        google::protobuf::FieldDescriptor::CPPTYPE_MESSAGE) {
      continue;
    }
    // We do not support fallback values for repeated or map fields.
    if (field->is_map() || field->is_repeated()) {
      continue;
    }
    if (field->message_type()->full_name() !=
        intrinsic_proto::assets::v1::ResolvedDependency::descriptor()
            ->full_name()) {
      continue;
    }

    const google::protobuf::FieldOptions& options = field->options();
    if (options.HasExtension(intrinsic_proto::assets::field_metadata)) {
      const intrinsic_proto::assets::FieldMetadata& field_metadata =
          options.GetExtension(intrinsic_proto::assets::field_metadata);
      if (field_metadata.dependency()
              .skill_annotations()
              .has_fallback_manifest_dependency_key()) {
        const std::string fallback_manifest_dependency_key =
            field_metadata.dependency()
                .skill_annotations()
                .fallback_manifest_dependency_key();
        auto it = fallback_manifest_dependencies.find(
            fallback_manifest_dependency_key);
        if (it != fallback_manifest_dependencies.end()) {
          // The message might be a dynamic proto so we serialize and
          // deserialize for modification. Using DynamicCastMessage would be
          // cleaner but does not work (gives nullptr) and static casting
          // produces a segfault.
          google::protobuf::Message& message =
              *parameter.GetReflection()->MutableMessage(&parameter, field);
          std::string serialized = message.SerializeAsString();
          ::intrinsic_proto::assets::v1::ResolvedDependency resolved_dependency;
          if (!resolved_dependency.ParseFromString(serialized)) {
            return absl::InvalidArgumentError(
                "Failed to parse ResolvedDependency from message.");
          }
          if (resolved_dependency.name().empty()) {
            resolved_dependency.set_name(it->second);
          }
          serialized = resolved_dependency.SerializeAsString();
          if (!message.ParseFromString(serialized)) {
            return absl::InvalidArgumentError(
                "Failed to parse ResolvedDependency back into message.");
          }
        }
      }
    }
  }
  return absl::OkStatus();
}

absl::Status Resolver::ResolveParameterDependencies(
    google::protobuf::Message& parameter, executive::SkillAction skill_action,
    const google::protobuf::DescriptorPool* pool,
    google::protobuf::MessageFactory* message_factory) const {
  return ResolveParameterDependenciesWithFallback(parameter, skill_action, {},
                                                  pool, message_factory);
}

absl::Status Resolver::ResolveParameterDependenciesWithFallback(
    google::protobuf::Message& parameter, executive::SkillAction skill_action,
    const absl::flat_hash_map<std::string, std::string>&
        fallback_manifest_dependencies,
    const google::protobuf::DescriptorPool* pool,
    google::protobuf::MessageFactory* message_factory) const {
  INTR_RETURN_IF_ERROR(
      PopulateFallbackValues(parameter, fallback_manifest_dependencies));

  // Use the recursively walk message  utility to recursively resolve
  // dependencies in the parameter proto.
  return RecursivelyWalkMessage(
      parameter,
      [this, fallback_manifest_dependencies, skill_action, pool,
       message_factory](
          google::protobuf::Message& msg) -> absl::StatusOr<bool> {
        return this->ResolveDependency(msg, skill_action, pool,
                                       message_factory);
      });
}

}  // namespace assets
}  // namespace intrinsic

File: intrinsic/assets/dependencies/resolver.go
// Copyright 2026 Intrinsic Innovation LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package resolver provides functionalities for resolving dependencies of Assets.
package resolver

import (
	"fmt"
	"strings"

	"intrinsic/assets/idutils"
	"intrinsic/assets/interfaceutils"
	"intrinsic/util/proto/registryutil"
	"intrinsic/util/proto/walkmessages"

	log "github.com/golang/glog"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"

	fieldmetadatapb "intrinsic/assets/proto/field_metadata_go_proto"
	grpcconnectionpb "intrinsic/assets/proto/v1/grpc_connection_go_proto"
	resolveddependencypb "intrinsic/assets/proto/v1/resolved_dependency_go_proto"

	descriptorpb "google.golang.org/protobuf/types/descriptorpb"
	anypb "google.golang.org/protobuf/types/known/anypb"
)

var (
	resolvedDependencyDescriptor = (&resolveddependencypb.ResolvedDependency{}).ProtoReflect().Descriptor()
	anyDescriptor                = (&anypb.Any{}).ProtoReflect().Descriptor()
)

// messageProcessor is a function that takes a message and processes it.
type messageProcessor func(proto.Message) (proto.Message, error)

// unpackAny uses the given file descriptor set to unmarshal the Any proto.
func unpackAny(config *anypb.Any, fds *descriptorpb.FileDescriptorSet) (proto.Message, error) {
	types, err := registryutil.NewTypesFromFileDescriptorSet(fds)
	if err != nil {
		log.Errorf("failed to create types from file descriptor set: %v", err)
		return nil, err
	}

	msgType, err := types.FindMessageByName(config.MessageName())
	if err != nil {
		log.Errorf("failed to find message by name: %v", err)
		return nil, err
	}
	msg := msgType.New().Interface()
	if err := config.UnmarshalTo(msg); err != nil {
		log.Errorf("failed to unmarshal Any: %v", err)
		return nil, err
	}
	return msg, nil
}

// deterministicAnyNew is an variant of anypb.New that uses deterministic
// marshalling.  This is at least consistent within the same binary, but may
// not match other languages or future versions of golang.
//
// Having a deterministic output is important because we presume we are able to
// generate a deterministic resources chart.  A change in an annotation can
// trigger a pod restart.  We deliberately include the serialized config for a
// service instance in the context_id annotation to restart pods so they pick
// up new config.  A randomized output can trigger unintended restarts.
func deterministicAnyNew(inner proto.Message) (*anypb.Any, error) {
	msg := new(anypb.Any)
	if err := anypb.MarshalFrom(msg, inner, proto.MarshalOptions{Deterministic: true}); err != nil {
		return nil, err
	}
	return msg, nil
}

// Resolver resolves interface-based asset dependencies for a given config.
type Resolver struct {
	address        string
	instanceHeader string
}

// New creates a new Resolver.
//
// The `address` is the address of the asset instance or the address of the ingress.
// The `instanceHeaderName“ is the name of the gRPC metadata header that contains the instance name
// of the asset.
func New(address, instanceHeaderName string) *Resolver {
	return &Resolver{
		address:        address,
		instanceHeader: instanceHeaderName,
	}
}

// ResolveConfigDependencies resolves the asset dependencies specified in the given config, and
// returns a new resolved config. The input config is not modified.
//
// This method recursively walks the proto message and resolves dependencies for
// [ResolvedDependency] message type that are annotated with the dependency field metadata.
// This [ResolvedDependency] type may be a single field, a repeated field, or a map with the values
// of [ResolvedDependency] type.
//
// This method unpacks the top level Any proto from the specified file descriptor set. Nested Any
// protos are packed iff its descriptors are provided in the file descriptor set. No errors will
// be returned for missing descriptors of nested Anys.
//
// If either of the input arguments is nil, this method returns the config unchanged.
func (r *Resolver) ResolveConfigDependencies(config *anypb.Any, fds *descriptorpb.FileDescriptorSet) (*anypb.Any, error) {
	if config == nil || fds == nil {
		return config, nil
	}

	resolved, err := r.resolveDependencies(config, fds)
	if err != nil {
		return nil, err
	}
	if resolved == nil {
		return config, err
	}
	return resolved, nil
}

// resolveDependencies will resolve fields marked as a dependency and return a
// new, updated Any proto.  If there are no dependencies, it returns nil.
func (r *Resolver) resolveDependencies(config *anypb.Any, fds *descriptorpb.FileDescriptorSet) (*anypb.Any, error) {
	msg, err := unpackAny(config, fds)
	if err != nil {
		return nil, err
	}

	// Keep track of whether we change state within the proto.  If we don't, then
	// we can return nil to indicate to the caller to just use the original value.
	// This avoids introducing any changes whatsoever and the overhead of
	// remarshalling the proto.  Ideally we could simply return config back to
	// the caller here, however, config is sometimes a copy of the original
	// obtained by marshalling and unmarshalling the proto, since the original is
	// a dynamic proto, rather than strictly an Any proto.
	var touched bool
	resolvedMsg, err := walkmessages.Recursively(msg, func(msg proto.Message) (proto.Message, bool, error) {
		fields := msg.ProtoReflect().Descriptor().Fields()
		for i := 0; i < fields.Len(); i++ {
			field := fields.Get(i)
			if field.Kind() != protoreflect.MessageKind {
				continue
			}

			// Get the message descriptor of the field. If the field is a map, get the message
			// descriptor of the map value.
			var msgDesc protoreflect.MessageDescriptor
			if field.IsMap() {
				if field.MapValue().Kind() != protoreflect.MessageKind {
					continue
				}
				msgDesc = field.MapValue().Message()
			} else {
				msgDesc = field.Message()
			}

			if msgDesc.FullName() == anyDescriptor.FullName() {
				if err := processField(field, msg, func(m proto.Message) (proto.Message, error) {
					// The message is likely a dynamic proto, so we need to marshal and unmarshal.
					data, err := proto.Marshal(m)
					if err != nil {
						return nil, err
					}
					if len(data) == 0 {
						return nil, nil
					}
					anyMsg := &anypb.Any{}
					if err := proto.Unmarshal(data, anyMsg); err != nil {
						return nil, err
					}
					resolvedAnyMsg, err := r.resolveDependencies(anyMsg, fds)
					if err != nil {
						log.Errorf("failed to resolve nested Any proto with type URL %q: %v. Ignoring this Any proto.", anyMsg.GetTypeUrl(), err)
						return anyMsg, nil
					}
					if resolvedAnyMsg == nil {
						return m, nil
					}
					touched = true
					return resolvedAnyMsg, nil
				}); err != nil {
					return nil, false, err
				}
				// If no error, either the Any proto was processed and any dependencies were resolved
				// successfully, or there was a missing FDS. In either case, we skip the rest of the
				// processing for this field.
				continue
			}

			if msgDesc.Name() != resolvedDependencyDescriptor.Name() {
				continue
			}

			options := field.Options().(*descriptorpb.FieldOptions)
			fieldMetadata := proto.GetExtension(options, fieldmetadatapb.E_FieldMetadata).(*fieldmetadatapb.FieldMetadata)
			requires := fieldMetadata.GetDependency().GetRequires()
			if fieldMetadata.GetDependency() == nil {
				// Since there is no dependency annotation, we assume here that the user does not want to
				// treat this field as a dependency specification.
				continue
			}
			requiresObject := fieldMetadata.GetDependency().GetRequiresObject() != nil

			if err := processField(field, msg, func(m proto.Message) (proto.Message, error) {
				return r.resolveStub(m, requires, requiresObject)
			}); err != nil {
				return nil, false, err
			}
			touched = true
		}
		return msg, true, nil
	})
	if err != nil {
		return nil, err
	}

	if !touched {
		return nil, nil
	}

	return deterministicAnyNew(resolvedMsg)
}

// processField processes a field of a proto message with a given processing function.
//
// If the field is a list or map, the function is called for each element.
// If the field is a message, the function is called for the message.
func processField(field protoreflect.FieldDescriptor, parentMsg proto.Message, f messageProcessor) error {
	if field.IsList() {
		list := parentMsg.ProtoReflect().Get(field).List()
		for i := 0; i < list.Len(); i++ {
			if list.Get(i).Message().Interface() != nil {
				val, err := f(list.Get(i).Message().Interface())
				if err != nil {
					return err
				}
				if val != nil {
					list.Set(i, protoreflect.ValueOfMessage(val.ProtoReflect()))
				}
			}
		}
	} else if field.IsMap() {
		m := parentMsg.ProtoReflect().Get(field).Map()
		var processErr error
		m.Range(func(key protoreflect.MapKey, value protoreflect.Value) bool {
			if value.Message().Interface() != nil {
				val, err := f(value.Message().Interface())
				if err != nil {
					processErr = err
					return false
				}
				if val != nil {
					m.Set(key, protoreflect.ValueOfMessage(val.ProtoReflect()))
				}
			}
			return true
		})
		if processErr != nil {
			return processErr
		}
	} else {
		if parentMsg.ProtoReflect().Has(field) {
			dynamicMsg := parentMsg.ProtoReflect().Get(field).Message().Interface()
			val, err := f(dynamicMsg)
			if err != nil {
				return err
			}
			if val != nil {
				parentMsg.ProtoReflect().Set(field, protoreflect.ValueOfMessage(val.ProtoReflect()))
			}
		}
	}
	return nil
}

// resolveStub returns a [ResolvedDependency] message with the interfaces field populated and the
// name field cleared.
func (r *Resolver) resolveStub(msg proto.Message, requires []string, requiresObject bool) (*resolveddependencypb.ResolvedDependency, error) {
	// The message is likely a dynamic proto, so we need to marshal and unmarshal.
	data, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return nil, nil
	}
	val := &resolveddependencypb.ResolvedDependency{}
	if err := proto.Unmarshal(data, val); err != nil {
		return nil, err
	}
	dependencyName := val.GetName()
	if dependencyName == "" {
		return val, nil
	}
	if len(val.GetInterfaces()) == 0 && len(requires) > 0 {
		val.Interfaces = make(map[string]*resolveddependencypb.ResolvedDependency_Interface, len(requires))
	}
	for _, require := range requires {
		if strings.HasPrefix(require, interfaceutils.GRPCURIPrefix) {
			val.Interfaces[require] = &resolveddependencypb.ResolvedDependency_Interface{
				Protocol: &resolveddependencypb.ResolvedDependency_Interface_Grpc_{
					Grpc: &resolveddependencypb.ResolvedDependency_Interface_Grpc{
						Connection: &grpcconnectionpb.GrpcConnection{
							Address: r.address,
							Metadata: []*grpcconnectionpb.GrpcConnection_Metadata{
								{
									Key:   r.instanceHeader,
									Value: dependencyName,
								},
							},
						},
					},
				},
			}
		} else if strings.HasPrefix(require, interfaceutils.DataURIPrefix) {
			if id, err := idutils.IDProtoFromString(dependencyName); err != nil {
				log.Errorf("failed to create data asset ID proto from dependency name: %v", err)
			} else {
				val.Interfaces[require] = &resolveddependencypb.ResolvedDependency_Interface{
					Protocol: &resolveddependencypb.ResolvedDependency_Interface_Data_{
						Data: &resolveddependencypb.ResolvedDependency_Interface_Data{
							Id: id,
						},
					},
				}
			}
		} else {
			return nil, fmt.Errorf("unsupported dependency type: %v", require)
		}
	}

	if requiresObject {
		val.Object = &resolveddependencypb.ResolvedDependency_Object{
			Name: dependencyName,
		}
	}

	// Clear the name field, as we have extracted all needed information.
	val.Name = ""

	return val, nil
}
